From 561a9fd3abf044b5dfafd4852097c1bccc8e98bb Mon Sep 17 00:00:00 2001
From: Dominique Martinet <dominique.martinet@atmark-techno.com>
Date: Mon, 1 Jul 2024 18:19:12 +0900
Subject: [PATCH] allow using a different sshd-session flavor

---
 Makefile.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.in b/Makefile.in
index e1b77ebc6495..df745d80863b 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -24,7 +24,7 @@ SSH_PROGRAM=@bindir@/ssh
 ASKPASS_PROGRAM=$(libexecdir)/ssh-askpass
 SFTP_SERVER=$(libexecdir)/sftp-server
 SSH_KEYSIGN=$(libexecdir)/ssh-keysign
-SSHD_SESSION=$(libexecdir)/sshd-session
+SSHD_SESSION=$(libexecdir)/sshd-session$(SSHD_SESSION_FLAVOR)
 SSH_PKCS11_HELPER=$(libexecdir)/ssh-pkcs11-helper
 SSH_SK_HELPER=$(libexecdir)/ssh-sk-helper
 PRIVSEP_PATH=@PRIVSEP_PATH@
--
2.39.2

