--- openssh-6.8p1.orig/loginrec.c	2015-03-17 07:49:20.000000000 -0200
+++ openssh-6.8p1/loginrec.c	2015-03-19 10:45:45.780276841 -0200
@@ -783,7 +783,7 @@
 	if (li->hostaddr.sa.sa_family == AF_INET)
 		utx->ut_addr = li->hostaddr.sa_in.sin_addr.s_addr;
 # endif
-# ifdef HAVE_ADDR_V6_IN_UTMP
+# ifdef HAVE_ADDR_V6_IN_UTMPX
 	/* this is just a 128-bit IPv6 address */
 	if (li->hostaddr.sa.sa_family == AF_INET6) {
 		sa6 = ((struct sockaddr_in6 *)&li->hostaddr.sa);
